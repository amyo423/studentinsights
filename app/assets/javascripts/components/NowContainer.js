import React from 'react';

// A container for testing that allows passing in a `nowFn`
// that will be shared with child components through context.
export default class NowContainer extends React.Component {
  getChildContext() {
    const {nowFn} = this.props;
    return {
      nowFn() { return nowFn(); }
    };
  }

  render() {
    const {children} = this.props;

    return children;
  }
}
NowContainer.childContextTypes = {
  nowFn: React.PropTypes.func
};
NowContainer.propTypes = {
  children: React.PropTypes.node.isRequired,
  nowFn: React.PropTypes.func.isRequired
};