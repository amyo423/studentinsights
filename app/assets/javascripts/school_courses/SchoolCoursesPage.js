import React from 'react';
import _ from 'lodash';
import GenericLoader from '../components/GenericLoader';
import School from '../components/School';
import Section from '../components/Section';
import SectionHeading from '../components/SectionHeading';
import ExperimentalBanner from '../components/ExperimentalBanner';
import {apiFetchJson} from '../helpers/apiFetchJson';
import {toMomentFromTime} from '../helpers/toMoment';

/*
Showing info about courses offered at a school.
*/
class SchoolCoursesPage extends React.Component {
  constructor(props) {
    super(props);
    this.fetchCourses = this.fetchCourses.bind(this);
    this.renderCourses = this.renderCourses.bind(this);
  }

  fetchCourses() {
    const {schoolId} = this.props;
    const url = `/api/schools/${schoolId}/courses`;
    return apiFetchJson(url);
  }

  render() {
    return (
      <div className="SchoolCoursesPage">
        <ExperimentalBanner />
        <GenericLoader
          promiseFn={this.fetchCourses}
          render={this.renderCourses} />
      </div>
    );
  }

  renderCourses(json) {
    const {courses, school} = json;
    const {nowFn} = this.context;
    const now = nowFn();

    const sortedCourses = _.sortBy(courses, course => course.course_number);
    return (
      <div style={styles.rendered}>
        <SectionHeading>
          Courses at <School id={school.id} name={school.name} style={{fontSize: 24, fontWeight: 300}}/>
        </SectionHeading>
        <table style={styles.table}>
          <thead>
            <tr>
              <th style={styles.cell}>Course</th>
              <th style={styles.cell}>Sections</th>
              <th style={styles.cell}>Total students</th>
              <th style={styles.cell}>Student<br />grade range</th>
              <th style={styles.cell}>Student<br />age range</th>
            </tr>
          </thead>
          <tbody>{sortedCourses.map(course => {
            const students = _.flatten(course.sections.map(s => s.students));
            const grades = _.sortBy(_.uniq(students.map(s => s.grade)));
            const ages =  _.sortBy(_.uniq(students.map(s => toMomentFromTime(s.date_of_birth).unix())).map(unix => {
              const birthdate = moment.unix(unix).utc();
              return now.clone().diff(birthdate, 'year');
            }));
            return (
              <tr key={course.id}>
                <td style={styles.cell}>{course.course_number} {course.course_description}</td>
                <td style={styles.cell}>{_.sortBy(course.sections, s => s.section_number).map(section =>
                  <Section
                    key={section.id}
                    id={section.id}
                    domain="https://somerville.studentinsights.org"
                    style={{display: 'block'}}
                    sectionNumber={section.section_number}
                    courseDescription={course.course_description} />
                )}</td>
                <td style={styles.cell}>{students.length}</td>
                <td style={styles.cell}>{grades.length > 1
                    ? <span>{_.first(grades)} - {_.last(grades)}</span>
                    : grades[0]}</td>
                <td style={styles.cell}>{_.first(ages)} - {_.last(ages)}</td>
              </tr>
            );
          })}</tbody>
        </table>
      </div>
    );
  }
}
SchoolCoursesPage.contextTypes = {
  nowFn: React.PropTypes.func.isRequired
};
SchoolCoursesPage.propTypes = {
  schoolId: React.PropTypes.string.isRequired
};


const styles = {
  rendered: {
    padding: 10
  },
  table: {
    margin: 10
  },
  cell: {
    textAlign: 'left',
    verticalAlign: 'top',
    padding: 5
  },
  raw: {
    fontFamily: 'monospace',
    fontSize: 10,
    padding: 10
  }
};


export default SchoolCoursesPage;