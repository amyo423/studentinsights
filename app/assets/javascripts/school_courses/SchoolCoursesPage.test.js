import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import SpecSugar from '../../../../spec/javascripts/support/spec_sugar';
import {withDefaultNowContext} from '../../../../spec/javascripts/support/NowContainer';
import fetchMock from 'fetch-mock/es5/client';
import SchoolCoursesPage, {SchoolCoursesPagePure} from './SchoolCoursesPage';
import schoolCoursesJson from '../../../../spec/javascripts/fixtures/schoolCoursesJson';


function testProps() {
  return {
    schoolId: '42'
  };
}


beforeEach(() => {
  fetchMock.restore();
  fetchMock.get('/api/schools/42/courses', schoolCoursesJson);
});

it('renders without crashing', () => {
  const props = testProps();
  const el = document.createElement('div');
  ReactDOM.render(withDefaultNowContext(<SchoolCoursesPage {...props} />), el);
});

SpecSugar.withTestEl('integration tests', container => {
  it('renders everything after fetch', done => {
    const props = testProps();
    const el = container.testEl;
    ReactDOM.render(withDefaultNowContext(<SchoolCoursesPage {...props} />), el);
    expect(el.innerHTML).toContain('This is an experimental prototype page!');

    setTimeout(() => {
      expect($(el).find('table tbody tr').length).toEqual(3);
      done();
    }, 0);
  });
});


describe('SchoolCoursesPagePure', () => {
  it('pure component matches snapshot', () => {
    const {courses, school} = schoolCoursesJson;
    const tree = renderer
      .create(withDefaultNowContext(<SchoolCoursesPagePure courses={courses} school={school} />))
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});

