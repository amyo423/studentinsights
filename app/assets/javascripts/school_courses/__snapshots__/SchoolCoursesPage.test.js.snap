// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SchoolCoursesPagePure pure component matches snapshot 1`] = `
<div
  style={
    Object {
      "padding": 10,
    }
  }
>
  <div
    className="SectionHeading"
    style={
      Object {
        "borderBottom": "1px solid #333",
        "padding": 10,
      }
    }
  >
    <h4
      style={
        Object {
          "color": "black",
          "display": "inline",
        }
      }
    >
      Courses at 
      <a
        className="School"
        href="/schools/9"
        style={
          Object {
            "fontSize": 24,
            "fontWeight": 300,
          }
        }
      >
        Localhost High
      </a>
    </h4>
  </div>
  <table
    style={
      Object {
        "border": "1px solid #eee",
        "borderCollapse": "collapse",
        "margin": 10,
      }
    }
  >
    <thead>
      <tr>
        <th
          style={
            Object {
              "backgroundColor": "#ccc",
              "border": "1px solid #eee",
              "fontWeight": "bold",
              "padding": 10,
              "textAlign": "left",
              "verticalAlign": "top",
            }
          }
        >
          Course
        </th>
        <th
          style={
            Object {
              "backgroundColor": "#ccc",
              "border": "1px solid #eee",
              "fontWeight": "bold",
              "padding": 10,
              "textAlign": "left",
              "verticalAlign": "top",
            }
          }
        >
          Sections
        </th>
        <th
          style={
            Object {
              "backgroundColor": "#ccc",
              "border": "1px solid #eee",
              "fontWeight": "bold",
              "padding": 10,
              "textAlign": "left",
              "verticalAlign": "top",
            }
          }
        >
          Total students
        </th>
        <th
          style={
            Object {
              "backgroundColor": "#ccc",
              "border": "1px solid #eee",
              "fontWeight": "bold",
              "padding": 10,
              "textAlign": "left",
              "verticalAlign": "top",
            }
          }
        >
          Student
          <br />
          grade range
        </th>
        <th
          style={
            Object {
              "backgroundColor": "#ccc",
              "border": "1px solid #eee",
              "fontWeight": "bold",
              "padding": 10,
              "textAlign": "left",
              "verticalAlign": "top",
            }
          }
        >
          Student
          <br />
          age range
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          style={
            Object {
              "border": "1px solid #eee",
              "padding": 10,
              "textAlign": "left",
              "verticalAlign": "top",
            }
          }
        >
          ART-302
           
          ART MAJOR FOUNDATIONS
        </td>
        <td
          style={
            Object {
              "border": "1px solid #eee",
              "padding": 10,
              "textAlign": "left",
              "verticalAlign": "top",
            }
          }
        >
          <a
            className="Section"
            href="/sections/3"
            style={
              Object {
                "display": "block",
              }
            }
          >
            ART MAJOR FOUNDATIONS
             (
            ART-302A
            )
          </a>
          <a
            className="Section"
            href="/sections/4"
            style={
              Object {
                "display": "block",
              }
            }
          >
            ART MAJOR FOUNDATIONS
             (
            ART-302B
            )
          </a>
        </td>
        <td
          style={
            Object {
              "border": "1px solid #eee",
              "padding": 10,
              "textAlign": "right",
              "verticalAlign": "top",
            }
          }
        >
          18
        </td>
        <td
          style={
            Object {
              "border": "1px solid #eee",
              "padding": 10,
              "textAlign": "right",
              "verticalAlign": "top",
            }
          }
        >
          -10
        </td>
        <td
          style={
            Object {
              "border": "1px solid #eee",
              "padding": 10,
              "textAlign": "right",
              "verticalAlign": "top",
            }
          }
        >
          13
           - 
          14
        </td>
      </tr>
      <tr>
        <td
          style={
            Object {
              "border": "1px solid #eee",
              "padding": 10,
              "textAlign": "left",
              "verticalAlign": "top",
            }
          }
        >
          BIO-700
           
          BIOLOGY 1 HONORS
        </td>
        <td
          style={
            Object {
              "border": "1px solid #eee",
              "padding": 10,
              "textAlign": "left",
              "verticalAlign": "top",
            }
          }
        >
          <a
            className="Section"
            href="/sections/2"
            style={
              Object {
                "display": "block",
              }
            }
          >
            BIOLOGY 1 HONORS
             (
            SHS-BIO-THUR
            )
          </a>
          <a
            className="Section"
            href="/sections/1"
            style={
              Object {
                "display": "block",
              }
            }
          >
            BIOLOGY 1 HONORS
             (
            SHS-BIO-TUES
            )
          </a>
        </td>
        <td
          style={
            Object {
              "border": "1px solid #eee",
              "padding": 10,
              "textAlign": "right",
              "verticalAlign": "top",
            }
          }
        >
          19
        </td>
        <td
          style={
            Object {
              "border": "1px solid #eee",
              "padding": 10,
              "textAlign": "right",
              "verticalAlign": "top",
            }
          }
        >
          <span>
            -10
             - 
            -9
          </span>
        </td>
        <td
          style={
            Object {
              "border": "1px solid #eee",
              "padding": 10,
              "textAlign": "right",
              "verticalAlign": "top",
            }
          }
        >
          13
           - 
          14
        </td>
      </tr>
      <tr>
        <td
          style={
            Object {
              "border": "1px solid #eee",
              "padding": 10,
              "textAlign": "left",
              "verticalAlign": "top",
            }
          }
        >
          SCI-201
           
          PHYSICS 1
        </td>
        <td
          style={
            Object {
              "border": "1px solid #eee",
              "padding": 10,
              "textAlign": "left",
              "verticalAlign": "top",
            }
          }
        >
          <a
            className="Section"
            href="/sections/5"
            style={
              Object {
                "display": "block",
              }
            }
          >
            PHYSICS 1
             (
            SCI-201A
            )
          </a>
          <a
            className="Section"
            href="/sections/6"
            style={
              Object {
                "display": "block",
              }
            }
          >
            PHYSICS 1
             (
            SCI-201B
            )
          </a>
        </td>
        <td
          style={
            Object {
              "border": "1px solid #eee",
              "padding": 10,
              "textAlign": "right",
              "verticalAlign": "top",
            }
          }
        >
          18
        </td>
        <td
          style={
            Object {
              "border": "1px solid #eee",
              "padding": 10,
              "textAlign": "right",
              "verticalAlign": "top",
            }
          }
        >
          -10
        </td>
        <td
          style={
            Object {
              "border": "1px solid #eee",
              "padding": 10,
              "textAlign": "right",
              "verticalAlign": "top",
            }
          }
        >
          13
           - 
          14
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
